pkg/controller/controllerutil/controllerutil.go

```diff
@@ -168,3 +170,39 @@ func mutate(f MutateFn, key client.ObjectKey, obj runtime.Object) error {
 
 // MutateFn is a function which mutates the existing object into it's desired state.
 type MutateFn func() error
+
+// AddFinalizer accepts a metav1 object and adds the provided finalizer if not present.
+func AddFinalizer(o metav1.Object, finalizer string) {
+	finalizers := sets.NewString(o.GetFinalizers()...)
```

Comment:
I hate to go back an forth here, but this was actually probably fine with the linear search -- namely, the list of finalizers is almost always tiny, which means the linear search is probably fast, and we avoid what's actually probably more "work" plus an allocation building the set (plus, we don't re-order the list of finalizers).  FWIW, k8s uses a normal linear search like the old code -- look for NeedsLBFinalizer and the finalizer code in the volume controller.

pkg/util/slice in apimachinery has a RemoveString and ContainsString that we can use to still make this simple to look at.

Reply:
makes sense. I noticed that most controller implementations (including the function you mentioned) make a DeepCopy to avoid mutating the informer cache. Is that something I should add here?

Reply:
err, i'm also not seeing those methods (or a pkg/util/slice in apimachinery). I think you mean these? https://github.com/kubernetes/kubernetes/blob/4d9e4f0b458c746f43836f558ebe61ec68bb7133/pkg/util/slice/slice.go

Happy to slice those out of k/k into apimachinery or somewhere else (pun heavily intended).

Reply:
I don't really understand the proper usage of controllers w.r.t informer cache because I saw this a lot pre-controller runtime, but don't see it as much in controllers using CR. I thought it might be under the hood somehow in CR (copying the object for you to avoid mutating the cache), but I didn't find it in internal/controller or in the manager code for injecting the cache. 

Would love to understand more deeply to make sure I'm doing the right thing.

Reply:
> I noticed most controller implementations make a DeepCopy

We automatically deepcopy before returning the object to you from the cache to avoid these sorts of issues: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/internal/cache_reader.go#L75

> I  think you mean these?

Yeah, those are the ones that I meant :-)

> CR (copying the object for you to avoid mutating the cache), but I didn't find it in internal/controller or in the manager code for injecting the cache.

See above :-)